web/interceptor/access_control.go

```diff
@@ -0,0 +1,181 @@
+	"GetCoursesByUser":        {user},
+	"UpdateUser":              {user, admin},
+	"GetEnrollmentsByUser":    {user, admin},
+	"GetSubmissions":          {student, group, teacher, courseAdmin},
+	"GetGroupByUserAndCourse": {group, teacher},
+	"CreateGroup":             {group, teacher},
+	"GetGroup":                {group, teacher},
+	"GetAssignments":          {student, teacher},
+	"GetEnrollmentsByCourse":  {student, teacher},
+	"GetRepositories":         {student, teacher},
+	"UpdateGroup":             {teacher},
+	"DeleteGroup":             {teacher},
+	"GetGroupsByCourse":       {teacher},
+	"UpdateCourse":            {teacher},
+	"UpdateEnrollments":       {teacher},
+	"UpdateAssignments":       {teacher},
+	"UpdateSubmission":        {teacher},
+	"UpdateSubmissions":       {teacher},
+	"RebuildSubmissions":      {teacher},
+	"CreateBenchmark":         {teacher},
+	"UpdateBenchmark":         {teacher},
+	"DeleteBenchmark":         {teacher},
+	"CreateCriterion":         {teacher},
+	"UpdateCriterion":         {teacher},
+	"DeleteCriterion":         {teacher},
+	"CreateReview":            {teacher},
+	"UpdateReview":            {teacher},
+	"GetReviewers":            {teacher},
+	"IsEmptyRepo":             {teacher},
+	"GetSubmissionsByCourse":  {courseAdmin},
+	"GetUserByCourse":         {teacher, admin},
+	"GetUsers":                {admin},
+	"GetOrganization":         {admin},
+	"CreateCourse":            {admin},
+}
+
+// AccessControl checks user information stored in the JWT claims agains the list of roles required to call the method.
+func AccessControl(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, request interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		req, ok := request.(requestID)
+		// The GetUserByCourse method sends a CourseUserRequest which has no IDs and needs a database query.
+		if !ok {
+			logger.Errorf("%s failed: message type '%s' does not implement IDFor interface",
+				method, reflect.TypeOf(request).String())
+			return nil, ErrAccessDenied
+		}
+		roles, ok := access[method]
+		if !ok {
+			logger.Errorf("No access roles defined for %s", method)
+			return nil, ErrAccessDenied
+		}
+		claims, err := tm.GetClaims(ctx)
+		if err != nil {
+			logger.Errorf("Access control: failed to get claims from request context: %v", err)
+			return handler(ctx, request)
+		}
+		for _, role := range roles {
+			switch role {
+			case none:
+				return handler(ctx, request)
+			case user:
+				if req.IDFor("user") == claims.UserID {
+					// Make sure the user is not updating own admin status.
+					if method == "UpdateUser" {
+						if req.(*qf.User).GetIsAdmin() != claims.Admin {
+							logger.Errorf("Access control: user %d attempted to change admin status from %v to %v",
+								claims.UserID, claims.Admin, req.(*qf.User).GetIsAdmin())
```

Comment:
Shouldn't there be a `return nil, ErrAccessDenied` here?

Reply:
Absolutely. Fixing, adding a test case for this specific situation.

Reply:
Fixed.